In [1]:
import pandas as pd
import numpy as np

import random

## QUESTION 17 -- CREATING A DATASET

In [71]:
random.seed(42)
# graph: x <- z_3 -> y, x -> z_1 <- y

z3 = [1 if random.random() > 0.6 else 0 for i in np.arange(0, 100000)] # 0 w/p 0.6, 1 w/p 0.4

# P(x = 1 | z_3 = 1) = 0.1 and P(x = 0 | z_3 = 1) = 0.9
# P(x = 1 | z_3 = 0) = 0.7 and P(x = 0 | z_3 = 0) = 0.3
x = [1 if (z3[i] == 1 and random.random() > 0.9) 
            or (z3[i] == 0 and random.random() > 0.3)
     else 0 for i in np.arange(0, 100000)]

# P(y = 1 | z_3 = 1) = 0.2 and P(y = 0 | z_3 = 1) = 0.8
# P(y = 1 | z_3 = 0) = 0.6 and P(y = 1 | z_3 = 0) = 0.4
y = [1 if (z3[i] == 1 and random.random() > 0.2) 
             or (z3[i] == 0 and random.random() > 0.4)
     else 0 for i in np.arange(0, 100000)]

# P(z_1 = 1 | y = 1, x = 1) = 0.5 and P(z_1 = 0 | y = 1, x = 1) = 0.5
# P(z_1 = 1 | y = 0, x = 1) = 0.3 and P(z_1 = 0 | y = 0, x = 1) = 0.7
# P(z_1 = 1 | y = 1, x = 0) = 0.2 and P(z_1 = 0 | y = 1, x = 0) = 0.8
# P(z_1 = 1 | y = 0, x = 0) = 0.25 and P(z_1 = 0 | y = 0, x = 0) = 0.75
z1 = [1 if (y[i] == 1 and x[i] == 1 and random.random() > 0.5) 
             or (y[i] == 0 and x[i] == 1 and random.random() > 0.7) 
             or (y[i] == 1 and x[i] == 0 and random.random() > 0.8) 
             or (y[i] == 0 and x[i] == 0 and random.random() > 0.75)
      else 0 for i in np.arange(0, 100000)]

In [72]:
df = pd.DataFrame({'z3':z3, 'x':x, 'y':y, 'z1':z1})
df.head()

z3  x  y  z1
0   1  0  1   1
1   0  1  0   0
2   0  1  1   0
3   0  1  1   1
4   1  0  0   1

In [73]:
## equation 42

# testing ONLY y=1, x=1, z_1=1 to validate problem
numerator = 0
denominator = 0
for z_3 in [0, 1]:
    prob1 = len(df[(df['y'] == 1) & (df['x'] == 1) & (df['z1'] == 1) & (df['z3'] == z_3)]) \
                / len(df[(df['x'] == 1) & (df['z1'] == 1) & (df['z3'] == z_3)])
    prob2 = len(df[(df['z1'] == 1) & (df['x'] == 1) & (df['z3'] == z_3)]) \
                / len(df[(df['x'] == 1) & (df['z3'] == z_3)])
    prob3 = len(df[df['z3'] == z_3]) / 100000
    numerator += (prob1 * prob2 * prob3)
    denominator += (prob2 * prob3)

numerator / denominator

0.7741729510446183

In [74]:
## equation 43

# y=1, x=1, z_1=1
prob1_2 = len(df[(df['z1'] == 1) & (df['x'] == 1) & (df['y'] == 1)]) \
            / len(df[(df['x'] == 1) & (df['y'] == 1)])
prob2_2 = len(df[df['y'] == 1]) / 100000

numerator2 = prob1_2 * prob2_2

numerator2 / denominator

0.7831534589452127

When n = 1000: equation 42 yielded approx 0.809 and equation 43 yielded approx 0.774.
Meanwhile, when n = 10000: equation 42 yielded approx 0.776 and equation 43 yielded approx 0.761.
Lastly, when n = 100000: equation 42 yielded approx 0.774 and equation 43 yielded approx 0.783.

Notice that as n increases, the similarity of the two equations increases as well!